session/utils.go

```diff
@@ -0,0 +1,223 @@
+			buf[pos] = '\\'
+			buf[pos+1] = '"'
+			pos += 2
+		case '\\':
+			buf[pos] = '\\'
+			buf[pos+1] = '\\'
+			pos += 2
+		default:
+			buf[pos] = c
+			pos++
+		}
+	}
+
+	return buf[:pos]
+}
+
+// EscapeSQL will escape input arguments into the sql string, doing necessary processing.
+// But it does not prevent you from doing EscapeSQL("select '%?", ";SQL injection!;") => "select '';SQL injection!;'".
+// It is still your responsibility to write safe SQL.
+func EscapeSQL(sql string, args ...interface{}) (string, error) {
+	holders := strings.Count(sql, "%?")
+	if holders == 0 {
+		return sql, nil
+	} else if holders > len(args) {
+		return "", errors.Errorf("missing arguments, have %d specifiers, and %d arguments", holders, len(args))
+	}
+	buf := make([]byte, 0, len(sql))
+	argPos := 0
+	for i := 0; i < len(sql); i += 2 {
+		q := strings.Index(sql[i:], "%?")
+		if q == -1 {
+			buf = append(buf, sql[i:]...)
+			break
+		}
+		buf = append(buf, sql[i:i+q]...)
+		i += q
+		arg := args[argPos]
+		argPos++
+
+		if arg == nil {
+			buf = append(buf, "NULL"...)
+			continue
+		}
+
+		switch v := arg.(type) {
+		case int:
+			buf = strconv.AppendInt(buf, int64(v), 10)
+		case int8:
+			buf = strconv.AppendInt(buf, int64(v), 10)
+		case int16:
+			buf = strconv.AppendInt(buf, int64(v), 10)
+		case int32:
+			buf = strconv.AppendInt(buf, int64(v), 10)
+		case int64:
+			buf = strconv.AppendInt(buf, v, 10)
+		case uint:
+			buf = strconv.AppendUint(buf, uint64(v), 10)
+		case uint8:
+			buf = strconv.AppendUint(buf, uint64(v), 10)
+		case uint16:
+			buf = strconv.AppendUint(buf, uint64(v), 10)
+		case uint32:
+			buf = strconv.AppendUint(buf, uint64(v), 10)
+		case uint64:
+			buf = strconv.AppendUint(buf, v, 10)
+		case float32:
+			buf = strconv.AppendFloat(buf, float64(v), 'g', -1, 32)
+		case float64:
+			buf = strconv.AppendFloat(buf, v, 'g', -1, 64)
+		case bool:
+			if v {
+				buf = append(buf, '1')
+			} else {
+				buf = append(buf, '0')
+			}
+		case time.Time:
+			if v.IsZero() {
+				buf = append(buf, "'0000-00-00'"...)
+			} else {
+				buf = append(buf, '\'')
+				buf = v.AppendFormat(buf, "2006-01-02 15:04:05")
+				buf = append(buf, '\'')
+			}
+		case json.RawMessage:
+			buf = append(buf, '\'')
+			buf = EscapeBytesBackslash(buf, v)
+			buf = append(buf, '\'')
+		case []byte:
+			if v == nil {
+				buf = append(buf, "NULL"...)
+			} else {
+				buf = append(buf, "_binary'"...)
+				buf = EscapeBytesBackslash(buf, v)
+				buf = append(buf, '\'')
+			}
+		case string:
+			buf = append(buf, '\'')
+			buf = EscapeStringBackslash(buf, v)
+			buf = append(buf, '\'')
+		case []string:
+			buf = append(buf, '(')
+			for i, k := range v {
+				if i > 0 {
+					buf = append(buf, ',')
+				}
+				buf = append(buf, '\'')
+				buf = EscapeStringBackslash(buf, k)
+				buf = append(buf, '\'')
+			}
+			buf = append(buf, ')')
+		default:
+			return "", errors.Errorf("unsupported %d-th argument: %v", argPos, arg)
+		}
+	}
+	return string(buf), nil
```

Comment:
The next time, you can play with [strings.Builder](https://pkg.go.dev/strings#Builder),  it avoids manipulate buffer manually and the stringer conversion.

Reply:
I know, it is a porting of go-sql-driver/mysql#interpolateParams. I did not bother much around the manipulation.